dace/transformation/dataflow/lift_einsum.py

```diff
@@ -0,0 +1,173 @@
+from dace.sdfg import utils as sdutil
+from dace.symbolic import pystr_to_symbolic
+from dace.transformation import transformation as xf
+
+
+class LiftEinsum(xf.SingleStateTransformation):
+    """
+    Detects a tensor operation that can be represented by an Einstein-notation sum (einsum, e.g., matrix
+    multiplication) and replaces the pattern with an ``Einsum`` library node.
+    """
+    EINSUM_CHARS = 'ijklmnopqrstuvwxyzabcdefgh'
+
+    map_entry = xf.PatternNode(nodes.MapEntry)
+    tasklet = xf.PatternNode(nodes.Tasklet)
+
+    @classmethod
+    def expressions(cls):
+        return [sdutil.node_path_graph(cls.map_entry, cls.tasklet)]
+
+    def can_be_applied(self, state: SDFGState, _, sdfg: SDFG, permissive=False):
+        contents = state.scope_subgraph(self.map_entry, False, False)
+        # Ensure map body contains only the tasklet
+        if len(contents.nodes()) != 1:
+            return False
+
+        # Check that indices match map indices
+        unique_chars = set()
+        input_chars = set()
+        output_chars = set()
+        for e in state.all_edges(self.tasklet):
+            memlet = e.data
+            if memlet.is_empty():
+                continue
+            if memlet.dynamic:
+                return False
+            if memlet.volume != 1 or memlet.subset.num_elements() != 1:
+                return False
+            ind = set(str(rb) for rb, _, _ in memlet.subset.ndrange())
+            unique_chars |= ind
+            if any(i != '0' and i not in self.map_entry.map.params for i in ind):
+                return False
+
+            # Keep track of input/output indices for WCR check
+            if e.dst is self.tasklet:
+                input_chars |= ind
+            else:
+                output_chars |= ind
+
+        if len(input_chars) == 0:
+            return False
+
+        # If there are too many characters for an einsum expression, fail
+        if len(unique_chars) == 0 or len(unique_chars) > len(self.EINSUM_CHARS):
+            return False
+
+        # Test outputs
+        oe = state.out_edges(self.tasklet)
+        if len(oe) != 1:
+            return False
+        oe = oe[0]
+        # Check for WCR if relevant
+        if input_chars - output_chars:
+            if oe.data.wcr is None or detect_reduction_type(oe.data.wcr) != dtypes.ReductionType.Sum:
+                return False
+        elif oe.data.wcr is not None:  # if, e.g., outer product, no WCR
+            return False
+
+        # Check tasklet contents
+        if self.tasklet.code.language != dtypes.Language.Python:
+            return False
+        if len(self.tasklet.code.code) != 1:
+            return False
+        expr = self.tasklet.code.code[0]
+        if not isinstance(expr, (ast.Assign, ast.AnnAssign)):
+            return False
+        if expr.value is None:
+            return False
+
+        try:
+            symexpr = pystr_to_symbolic(astutils.unparse(expr.value))
```

Comment:
Very interesting approach. I would have parsed the expression using the Python AST to verify that it is a sequence of scalar multiplications. Do you foresee any issues with either approach?

Reply:
I use the symbolic approach on purpose, in order to allow behavior that some frontends inject, (for example `1 * value` in DaCeML's auto-differentiation), as well as to detect the alpha coefficient and let sympy do some algebraic identities. The other approach might be more dangerous to write IMO.

Since we verify it is one expression (i.e., symbolic-conversion capable), it should always work when the tasklet is well-formulated.